pilot/pkg/config/clusterregistry/secretcontroller.go

```diff
@@ -244,31 +246,32 @@ func addMemberCluster(s *corev1.Secret, c *Controller) {
 				Controller:       kubectl,
 			})
 		stopCh := make(chan struct{})
+		c.cs.rc[key].ControlChannel = stopCh
 		_ = kubectl.AppendServiceHandler(func(*model.Service, model.Event) { c.discoveryService.ClearCache() })
 		_ = kubectl.AppendInstanceHandler(func(*model.ServiceInstance, model.Event) { c.discoveryService.ClearCache() })
 
 		go kubectl.Run(stopCh)
 	}
 	// TODO Add exporting a number of cluster to Prometheus
 	// for now for debbuging purposes, print it to the log.
-	log.Infof("Number of clusters in the cluster store: %d", len(c.cs.clientConfigs))
+	log.Infof("Number of clusters in the cluster store: %d", len(c.cs.rc))
 }
 
-func deleteMemberCluster(s string, cs *ClusterStore) {
-	cs.storeLock.Lock()
-	defer cs.storeLock.Unlock()
+func deleteMemberCluster(s Metadata, c *Controller) {
+	c.cs.storeLock.Lock()
+	defer c.cs.storeLock.Unlock()
 	// Check if there is a cluster member with the specified name
-	if _, ok := cs.clientConfigs[s]; ok {
+	if _, ok := c.cs.rc[s]; ok {
 		log.Infof("Deleting cluster member: %s", s)
-		delete(cs.clientConfigs, s)
-		for i, c := range cs.clusters {
-			if c.ObjectMeta.Name == s {
-				cs.clusters = append(cs.clusters[:i], cs.clusters[i+1:]...)
-				break
-			}
-		}
+		// Deleting Service registry associated with controller
+		c.serviceController.DeleteRegistry(GetClusterID(c.cs.rc[s].Cluster))
+		// Stop controller
+		close(c.cs.rc[s].ControlChannel)
+		<-c.cs.rc[s].ControlChannel
```

Comment:
can't receive after the channel has been closed.

Reply:
This is the normal practice of informing go routine to stop via stop channel and waiting until that go routine exits.